File: python/group_anagrams.py
from collections import defaultdict

class Solution:
    def make_hash_table(self, strs) -> Dict[str, List[str]]:
        result = defaultdict(list)
        for s in strs:
            ss = "".join(sorted(s))
            result[ss].append(s)
        return result

    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        ht = self.make_hash_table(strs)
        result = []
        strs = set(strs)
        while strs:
            current_string = strs.pop()
            css = "".join(sorted(current_string))
            anagrams = ht.pop(css, [current_string])
            for anagram in anagrams:
                try:
                    strs.remove(anagram)
                except KeyError:
                    pass
            result.append(anagrams)
        return result

File: rust/src/group_anagrams.rs
use std::collections::HashMap;

impl Solution {
    pub fn group_anagrams(strs: Vec<String>) -> Vec<Vec<String>> {
        let mut ht: HashMap<String, Vec<String>> = HashMap::new();
        for s in strs.iter() {
            let s_sorted = Solution::sort_string(s.to_string());
            let mut anagrams = ht.get_mut(&s_sorted);
            if anagrams.is_some() {
                anagrams.unwrap().push(s.to_string());
            } else {
                ht.insert(s_sorted, vec![s.to_string()]);
            }
        }
        let mut result = Vec::with_capacity(ht.len());
        for v in ht.values() {
            result.push(v.to_vec());
        }
        return result;
    }

    fn sort_string(s: String) -> String {
        let mut char_vec = s.chars().collect::<Vec<char>>();
        char_vec.sort();
        let s_sorted = char_vec.into_iter().collect::<String>();
        return s_sorted;
    }
}
